mmeds/util.py

```diff
@@ -1000,55 +1002,121 @@ def make_pheniqs_config(reads_forward, reads_reverse, barcodes_forward, barcodes
-                    if not code[0][i] == map_hash[sample][0][i]:
-                        diff += 1
-                # Compare reverse barcodes and count diff
-                for i in range(len(code[1])):
-                    if not code[1][i] == map_hash[sample][1][i]:
-                        diff += 1
-
-                # If diff does not exceed N, write read to output
-                if diff <= num_allowed_errors:
-                    out += line
-                    out += f.readline()
-                    out += f.readline()
-                    out += f.readline()
-                # Else skip read
-                else:
-                    f.readline()
-                    f.readline()
-                    f.readline()
+def strip_error_barcodes(num_allowed_errors,
+                         mapping_file,
+                         input_dir,
+                         output_dir,
+                         verbose,
+                         filename_template=fig.QIIME_FILENAME_TEMPLATE,
+                         sample_id_cats=fig.QIIME_SAMPLE_ID_CATS,
+                         forward_barcode_cats=fig.QIIME_FORWARD_BARCODE_CATS,
+                         reverse_barcode_cats=fig.QIIME_REVERSE_BARCODE_CATS):
+    """
+    Strip reads with errors from demultiplexed fastq files and write new file content
+    This has only been tested using input that has been demultiplexed using 'pheniqs'
+    =================================================================================
+    :num_allowed_errors: Maximum number of errors in barcode pairs to not be stripped
+    :mapping_file: File containing sample IDs and barcode pairs
+    :input_dir: Directory containing demultiplexed fastq files
+    :output_dir: Directory in which to store stripped fastq files
+    :verbose: Print information to stdout
+    :filename_template: String with formatting for sample id and (1 or 2) for forward/reverse
+    :sample_id_cats: tuple with sample id header text
+    :forward_barcode_cats: tuple with forward barcode header text
+    :reverse_barcode_cats: tuple with reverse barcode header text
+    """
+    # Read in mapping file
+    output_content = {}
+    map_df = pd.read_csv(Path(mapping_file), sep='\t', header=[0, 1], na_filter=False)
 
-                line = f.readline()
+    # Only three columns are needed
+    sample_ids = map_df[sample_id_cats[0]][sample_id_cats[1]]
+    forward_barcodes = map_df[forward_barcode_cats[0]][forward_barcode_cats[1]]
+    reverse_barcodes = map_df[reverse_barcode_cats[0]][reverse_barcode_cats[1]]
 
-            # Write output file to output_dir
-            p_out = Path(output_dir) / filename
-            p_out.touch()
-            p_out.write_bytes(gzip.compress(out.encode('utf-8')))
-            print(count, 'written', filename)
-            count += 1
+    # Hash data for easy access
+    map_hash = {key: (value1, value2) for (key, value1, value2) in zip(sample_ids, forward_barcodes, reverse_barcodes)}
```

Comment:
why not just glob the directory i.e. `Path.glob('*.fastq')` rather than doing the fastq filename template?

Reply:
Well that would also work, but I would have to add an extra checks to get the right file to match the current sample id, and it's that kind of check that caused the bug last time. With the template, we immediately get the right file

Reply:
I think it's better for the concern that Jose raised in config.py about having that template solution eventually break. You could always have a check that the file matches the template and then log if it isn't a match or something